components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3841,39 +3791,84 @@ subroutine ice_sedimentation(kts,kte,ktop,kbot,kdir,    &
 
 end subroutine ice_sedimentation
 
-
-subroutine generalized_sedimentation(kts, kte, kdir, k_temp, k_qxtop, dt_sub, inv_dzq, inv_rho, flux_qx, qx)
+subroutine generalized_sedimentation(kts, kte, kdir, k_qxtop, k_qxbot, kbot, Co_max, dt_left, prt_accum, inv_dzq, inv_rho, rho, &
+     num_arrays, vs, fluxes, qnx)
 
    implicit none
 
-   integer, intent(in) :: kts, kte, kdir, k_temp, k_qxtop
-   real(rtype), intent(in) :: dt_sub
+   integer, intent(in) :: kts, kte, kdir, k_qxtop, kbot, num_arrays
+   integer, intent(inout) :: k_qxbot
+   real(rtype), intent(in) :: Co_max
+   real(rtype), intent(inout) :: dt_left, prt_accum
    real(rtype), dimension(kts:kte), intent(in) :: inv_dzq
    real(rtype), dimension(kts:kte), intent(in) :: inv_rho
-   real(rtype), dimension(kts:kte), intent(in) :: flux_qx
-   real(rtype), dimension(kts:kte), intent(inout) :: qx
+   real(rtype), dimension(kts:kte), intent(in) :: rho
 
-   integer :: k
-   real(rtype) :: fluxdiv_qx
+   type(realptr), intent(in), dimension(num_arrays) :: vs, fluxes, qnx
 
-   !--- for top level only (since flux is 0 above)
-   k = k_qxtop
-   !- compute flux divergence
-   fluxdiv_qx = -flux_qx(k)*inv_dzq(k)
-   !- update prognostic variables
-   qx(k) = qx(k) + fluxdiv_qx*dt_sub*inv_rho(k)
+   integer :: tmpint1, k_temp, k, i
+   real(rtype) :: dt_sub
 
-   do k = k_qxtop-kdir,k_temp,-kdir
-      !-- compute flux divergence
-      fluxdiv_qx = (flux_qx(k+kdir) - flux_qx(k))*inv_dzq(k)
-      !-- update prognostic variables
-      qx(k) = qx(k) + fluxdiv_qx*dt_sub*inv_rho(k)
+   !-- compute dt_sub
+   tmpint1 = int(Co_max+1._rtype)    !number of substeps remaining if dt_sub were constant
+   dt_sub  = min(dt_left, dt_left/float(tmpint1))
 
-   end do
+   ! -- Move bottom cell down by 1 if not at ground already
+   if (k_qxbot.eq.kbot) then
+      k_temp = k_qxbot
+   else
+      k_temp = k_qxbot-kdir
+   endif
+
+   call calc_first_order_upwind_step(kts, kte, kdir, k_temp, k_qxtop, dt_sub, rho, inv_rho, inv_dzq, num_arrays, fluxes, vs, qnx)
+
+   !accumulated precip during time step
+   if (k_qxbot.eq.kbot) prt_accum = prt_accum + fluxes(1)%p(kbot)*dt_sub
 
+   dt_left = dt_left - dt_sub  !update time remaining for sedimentation
+   if (k_qxbot.ne.kbot) k_qxbot = k_qxbot - kdir
 
 end subroutine generalized_sedimentation
 
+subroutine calc_first_order_upwind_step(kts, kte, kdir, kbot, k_qxtop, dt_sub, rho, inv_rho, inv_dzq, num_arrays, fluxes, vs, qnx)
+
+  implicit none
+
+  integer, intent(in) :: kts, kte, kdir, kbot, k_qxtop, num_arrays
+  real(rtype), intent(in) :: dt_sub
+  real(rtype), dimension(kts:kte), intent(in) :: rho, inv_rho, inv_dzq
+  type(realptr), intent(in), dimension(num_arrays) :: fluxes, vs, qnx
```

Comment:
I'm assuming that if you pass a pointer array it doesn't need to be designated as an "out" field if you make changes inside it?

Reply:
I'm definitely pushing beyond my comfort zone with fortran in this PR, so I'm not sure this is correct. That said, all our test platforms accepted this as-is. The "in" appears to apply just to the pointers and not their contents.

Reply:
I am likewise untested using pointer arrays in Fortran.  I wanted to check just in case, but if it is B4B then it must be true that the qnx values inside the pointer array are being changed.  It would make sense in one way that the pointer doesn't change since it is a location in memory and not the actual memory value itself.